pennylane/gradients/jvp.py

```diff
@@ -89,46 +74,91 @@ def compute_jvp_single(tangent, jac):
+    #                                              tensor return type and k mixed params
+    #
+    # Note that the return type dimensions always remain the same, whereas the
+    # dimensions of the tensor parameters may differ.
+    #
+    # In this scenario, another case separation is used: If any of the parameters is a
+    # tensor (i.e. not a scalar), all tangent entries are reshaped into one-dimensional
+    # tensors with shapes (prod_l_i,) and then stacked to one one-dimensional tensor.
+    # If there are no tensor parameters, the tangent is just stacked and reshaped.
+    # The Jacobians are reshaped to have the dimensions (r_1, ... r_R, prod_l_i)
+    # and then are concatenated along their last (potentially mismatching) axis.
+    # This is followed by a tensordot contraction over the axes of size sum_i prod_l_i.
+
     if jac is None:
         return None
-
-    tangent = qml.math.stack(tangent)
-    jac = _convert(jac, tangent)
-
-    # Single param
-    if not isinstance(jac, tuple):
+    single_param = not isinstance(jac, tuple)
+    if (single_param and jac.shape == (0,)) or (not single_param and len(jac) == 0):
         # No trainable parameters
-        if jac.shape == (0,):
-            res = qml.math.zeros((1, 0))
-            return res
-
-        tangent = qml.math.reshape(tangent, (1,))
-
-        # No dimension e.g. expval
-        if jac.shape == ():
-            jac = qml.math.reshape(jac, (1,))
-
-        # With dimension e.g. probs
-        else:
-            jac = qml.math.reshape(jac, (1, -1))
-        res = qml.math.tensordot(jac, tangent, [[0], [0]])
-    # Multiple params
+        return qml.math.zeros((1, 0))
+
+    if single_param:
+        tangent = qml.math.stack(tangent)
+        num_l = len(tangent.shape[1:])
+        tangent = qml.math.reshape(tangent, (-1,))
+        prod_l = tangent.shape[0]
+        new_shape = (jac.shape if num_l == 0 else jac.shape[:-num_l]) + (prod_l,)
+        jac = qml.math.cast(qml.math.convert_like(jac, tangent), tangent.dtype)
+        jac = qml.math.reshape(jac, new_shape)
+        return qml.math.tensordot(jac, tangent, [[-1], [0]])
+
+    if isinstance(tangent, (tuple, list)) and any(t.ndim > 0 for t in tangent):
+        num_l = [t.ndim for t in tangent]
+        tangent = [qml.math.reshape(t, (-1,)) for t in tangent]
+        prod_l = [t.shape[0] for t in tangent]
+        tangent = qml.math.hstack(tangent)
     else:
-        # No trainable parameters (adjoint)
-        if len(jac) == 0:
-            res = qml.math.zeros((1, 0))
-            return res
-
-        jac = qml.math.stack(jac)
-
-        # No dimension e.g. expval
-        if jac[0].shape == ():
-            res = qml.math.tensordot(jac, tangent, 1)
-
-        # With dimension e.g. probs
-        else:
-            res = qml.math.tensordot(jac, tangent, [[0], [0]])
-    return res
+        num_l = [0] * len(tangent)
+        prod_l = [1] * len(tangent)
+        tangent = qml.math.stack(tangent)
```

Comment:
Why this distinction? 

FYI: We can move the 1st and 3rd lines outside of the `if`/`else` statement

Reply:
I think this distinction was left over from a way more complicated version of the code that I had inbetween.
So I think this might indeed be unnecessary. One question though: In the `else` clause we have `tangent` be a builtin sequence of scalars, which all would get reshaped to have shape `(1,)`. Do we want to do that extra work? I imagine this to be a very common use case of the function.

Reply:
I tried to make it work without the case distinction (note the `hstack` vs `stack`) but did not find a nice solution.
Happy about any suggestions you might have :)

Reply:
Ah ok! No worries we can leave it like this 👍